Improve README, Align API and improve defaults

rhrazdil opened the pull request:
Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>

- Improve documentation of the plugin API provided in README.md file
- provide more descriptive examples

After addressing Peters review, the following changes are included in this PR:
- improve defaulting
  - when `ingress.subnets` is not specified or empty array, do not allow any subnet
  - when `ingress.ports` is not specified or empty array, do not allow any port
  - to allow all subnets, do not specify `ingress.subnets[].cidr` or set it to empty string
  - to allow all ports, do not specify `ingress.ports[].port` or set it to empty string, the similar can be done for protocol. Any combination is allowed. If both `port` and `protocol` felds are not specified, all ports on all protocols are allowed.
- Filter traffic per-policy. Meaning that in order to allow a packet, it must match both, port and subnet within that policy.
  in other words, it's not possible to allow subnet in one policy and a port in another policy.

Commit:
Improve README

Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>
Co-authored-by: Miguel Duarte Barroso <mdbarroso@redhat.com>
Co-authored-by: Petr Horacek <p.horacek94@gmail.com>

phoracek commented on README.md:
```diff
@@ -10,18 +13,15 @@ This plugin allows user to define allow list for ingress and egress packet filte
-Each configMap additionally needs to have "cidr-filtering-cni" label.
-
-See the example below:
+When using the cidr-filtering-cni plugin, all ingress and egress traffic is dropped by default. To enable cidr-filtering-cni plugin, include it in the
```
Could we explain how to install it first?

rhrazdil replied:
Done

phoracek commented on README.md:
```diff
@@ -1,6 +1,9 @@
 # cidr-filtering-cni
 
-This plugin allows user to define allow list for ingress and egress packet filtering for a Kubernetes pod.
+This CNI plugin isolates pods from external endpoints. By default, all ingress and egress traffic is denied.
```
Does this work for regular Pods too (with IPAM assigning an IP to an interface)? I think we should mention that it is specific for filtering on bridges created in the Pod (if that is the case), and useful for kubevirt. So people don't confuse it with something useful for Pods too

maiqueb replied:
+1. Be sure to mention this is only useful if your pod interface is a port of a bridge.

maiqueb replied:
Another good point here - I think this solution as is does not work when IPAM is involved.

... it could, but in a hacky way: if you see an IP address in the result of the previous CNI, you can modify your iface name to `ifaceName += ifaceName + "-nic"`

We'd have to monitor that swticharoo on kubevirt.

Thoughts ?

rhrazdil replied:
IIUC, we don't need to support IPAM right of the bat, as it's not supported on d/s either

maiqueb replied:
Fair enough. That makes it legit to hold off on this.

phoracek commented on README.md:
```diff
@@ -1,6 +1,9 @@
 
-This plugin allows user to define allow list for ingress and egress packet filtering for a Kubernetes pod.
+This CNI plugin isolates pods from external endpoints. By default, all ingress and egress traffic is denied.
+Users may create special configMaps in the pod namespace to indicate allowed ingress or egress connections.
```
```suggestion
Users may create special ConfigMaps in the pod namespace to indicate allowed ingress or egress connections.
```

phoracek commented on README.md:
```diff
@@ -10,18 +13,15 @@ This plugin allows user to define allow list for ingress and egress packet filte
-
-See the example below:
+When using the cidr-filtering-cni plugin, all ingress and egress traffic is dropped by default. To enable cidr-filtering-cni plugin, include it in the
+NetworkAttachment definition. See the following example:
```
```suggestion
NetworkAttachmentDefinition. See the following example:
```

phoracek commented on README.md:
```diff
@@ -36,80 +36,172 @@ spec:
+is the `cidr-filtering-cni`, that handles configuration of nftables rules on the pod.
+
+By default, all ingress and egress traffic is dropped.
+To allow specific ingress and egress CIDR blocks/ports, create configMaps(s) referencing a NetworkAttachmentDefinition spec.config name in the configMaps label.
```
```suggestion
To allow specific ingress and egress CIDR blocks/ports, create ConfigMaps(s) referencing a NetworkAttachmentDefinition spec.config name in the ConfigMap's label.
```

phoracek commented on README.md:
```diff
@@ -36,80 +36,172 @@ spec:
+
+By default, all ingress and egress traffic is dropped.
+To allow specific ingress and egress CIDR blocks/ports, create configMaps(s) referencing a NetworkAttachmentDefinition spec.config name in the configMaps label.
+Each configMap additionally needs to have `cidr-filtering-cni` label to be enabled for the cidr-filtering-cni plugin.
```
```suggestion
Each ConfigMap additionally needs to have `cidr-filtering-cni` label to be enabled for the cidr-filtering-cni plugin.
```

phoracek commented on README.md:
````diff
@@ -36,80 +36,172 @@ spec:
+```
+
+The `example-network` specifies two CNI plugins, the first cni in the above is a `cnv-bridge`, connecting pods to linux-bridge `br1`, and the second
+is the `cidr-filtering-cni`, that handles configuration of nftables rules on the pod.
````
`cni` in the binary name is redundant

phoracek commented on README.md:
```diff
@@ -1,6 +1,9 @@
 # cidr-filtering-cni
```
It is not limited to CIDR. Perhaps bridge-filtering? If the CNI is special in filtering in the bridge table. Or maybe even bridge-binding-filtering?

@maiqueb thoughts?

maiqueb replied:
You raise valid points. I agree we should send the message this is only useful when the interface to which this applies to is a bridge port.

I could go for `bridge-filtering`.

maiqueb commented on README.md:
```diff
@@ -1,6 +1,9 @@
+This CNI plugin isolates pods from external endpoints. By default, all ingress and egress traffic is denied.
+Users may create special configMaps in the pod namespace to indicate allowed ingress or egress connections.
+This can be done in both directions - ingress and egress - for network layers L3 and L4.
+The supperted L4 protocols are UDP and TCP.
```
```suggestion
The supported L4 protocols are UDP and TCP.
```

phoracek commented on README.md:
```diff
@@ -36,80 +36,172 @@ spec:
+To allow specific ingress and egress CIDR blocks/ports, create configMaps(s) referencing a NetworkAttachmentDefinition spec.config name in the configMaps label.
+Each configMap additionally needs to have `cidr-filtering-cni` label to be enabled for the cidr-filtering-cni plugin.
+
+The following manifest is an example of a minimal configMap, that doesn't yet allow any traffic. It's given as an example, since denying all traffic is done by default, it's
```
```suggestion
The following manifest is an example of a minimal ConfigMap, that doesn't yet allow any traffic. It's given as an example, since denying all traffic is done by default, it's
```

maiqueb replied:
```suggestion
The following manifest is an example of a minimal `configMap`, that doesn't yet allow any traffic. It's given as an example, since denying all traffic is done by default, it's
```

phoracek commented on README.md:
```diff
@@ -36,80 +36,172 @@ spec:
+Notice the labels of the configMap.
+The first label, `cidr-filtering-cni` ensures that the configMap configuration is collected by the plugin, as
+the plugin only lists configMaps with this label.
+The second label, `br1-with-cidr-filtering` refers to the NetworkAttachmentDefinition spec name, in the cni configuration JSON.
```
```suggestion
Notice the labels of the ConfigMap.
The first label, `cidr-filtering-cni` ensures that the ConfigMap configuration is collected by the plugin, as
the plugin only lists configMaps with this label.
The second label, `br1-with-cidr-filtering` refers to the NetworkAttachmentDefinition spec name, in the CNI configuration JSON.
```

phoracek commented on README.md:
```diff
@@ -36,80 +36,172 @@ spec:
+The second label, `br1-with-cidr-filtering` refers to the NetworkAttachmentDefinition spec name, in the cni configuration JSON.
+
+
+> **_NOTE:_**  A configMap may refer to many NetworkAttachmentDefinition specs. The order in which configMaps are process is not defined.
```
```suggestion
> **_NOTE:_**  A ConfigMap may refer to many NetworkAttachmentDefinition specs. The order in which ConfigMaps are process is not defined.
```

phoracek commented on README.md:
```diff
@@ -36,80 +36,172 @@ spec:
+
+
+To allow a pod to communicate to external endpoints, let's create the following configMap, that allows pod to reach all ports
+on a local network:
```
```suggestion
To allow a pod to communicate to external endpoints, let's create the following ConfigMap, that allows pod to reach all ports
on a local network:
```

phoracek commented on README.md:
````diff
@@ -36,80 +36,172 @@ spec:
+    }
+```
+
+The `egress.subnets` attribute allows pod to reach to any IP address in subnet `192.168.0.0/16`, except for subnet `192.168.150.0/24` and `192.168.151.151`.
````
Is `except` available in NetworkPolicy API too? If not, we should hide it from the README

rhrazdil replied:
It is

phoracek commented on README.md:
````diff
@@ -36,80 +36,172 @@ spec:
+```
+
+The `egress.subnets` attribute allows pod to reach to any IP address in subnet `192.168.0.0/16`, except for subnet `192.168.150.0/24` and `192.168.151.151`.
+The `egress.ports` attribute allows pod to reach ports 80, and 8080 over TCP protocol.
````
Does this mean that destination bust be one of the ports on the subnet?

rhrazdil replied:
No, we're not enforcing that.

rhrazdil replied:
We are now (with this PR)

phoracek commented on README.md:
```diff
@@ -36,80 +36,172 @@ spec:
+The `egress.subnets` attribute allows pod to reach to any IP address in subnet `192.168.0.0/16`, except for subnet `192.168.150.0/24` and `192.168.151.151`.
+The `egress.ports` attribute allows pod to reach ports 80, and 8080 over TCP protocol.
+
+To allow any webserver responses to be allowed to reach the pod, create the following configmap:
```
```suggestion
To allow any webserver responses to be allowed to reach the pod, create the following ConfigMap:
```

maiqueb commented on README.md:
```diff
@@ -1,6 +1,9 @@
 # cidr-filtering-cni
 
-This plugin allows user to define allow list for ingress and egress packet filtering for a Kubernetes pod.
+This CNI plugin isolates pods from external endpoints. By default, all ingress and egress traffic is denied.
```
I would try to avoid using terms such as endpoints (which have a meaning of its own on kubernetes), and also external (external to what ?... the cluster ?.. the pod ? ... ).

How about something like: "This CNI plugin allows the user to control traffic flow at the IP address or port level (OSI layer 3 or 4) for particular applications in the Kubernetes cluster, thus specifying how a [pod](https://kubernetes.io/docs/concepts/workloads/pods/) is allowed to communicate with various network "entities".

maiqueb commented on README.md:
```diff
@@ -1,6 +1,9 @@
 
-This plugin allows user to define allow list for ingress and egress packet filtering for a Kubernetes pod.
+This CNI plugin isolates pods from external endpoints. By default, all ingress and egress traffic is denied.
+Users may create special configMaps in the pod namespace to indicate allowed ingress or egress connections.
+This can be done in both directions - ingress and egress - for network layers L3 and L4.
```
ingress / egress feels somewhat redundant - they appear twice in consecutive sentences.

maiqueb commented on README.md:
```diff
@@ -36,80 +36,172 @@ spec:
+Each configMap additionally needs to have `cidr-filtering-cni` label to be enabled for the cidr-filtering-cni plugin.
+
+The following manifest is an example of a minimal configMap, that doesn't yet allow any traffic. It's given as an example, since denying all traffic is done by default, it's
+not needed to be explicitly created.
```
*it isn't* reads better than *it's not* imo.

maiqueb commented on README.md:
```diff
@@ -36,80 +36,172 @@ spec:
+> **_NOTE:_**  A configMap may refer to many NetworkAttachmentDefinition specs. The order in which configMaps are process is not defined.
+
+
+To allow a pod to communicate to external endpoints, let's create the following configMap, that allows pod to reach all ports
```
```suggestion
To allow a pod to communicate to external entities, let's create the following configMap, that allows pod to reach all ports
```

maiqueb commented on README.md:
```diff
@@ -36,80 +36,172 @@ spec:
-allows all ports.
-
-To do the opposite - not allow any port - leave the `ports` attribute unspecified. The behaviour for subnets works analogically.
+When using a secondary network where clients obtain IP addresses from a DHCP server, allow the `255.255.255.255` IP
+address so that clients are allowed to send request to this address.
```
Really good point; you could be more specific though:

```suggestion
When using a secondary network where clients obtain IP addresses from a DHCP server, users must allow egress to  the `255.255.255.255` IP address, thus allowing the `DHCP Discover` message to be sent.
```

maiqueb reviewed (changes requested):
Probably it makes sense to highlight the consequence of us implementing this directly in the CNI layer - it will not react to updates of the configuration. 

The pod must be re-scheduled for the new rule configuration to apply.

phoracek commented on README.md:
```diff
@@ -36,80 +36,172 @@ spec:
           }
-        ]
+        ],
+        "ports": []
```
This is the default, no?

rhrazdil replied:
This allows all ports, which is not default.
Default is denying all, which is the same as not specifying the ports key at all.
This is the same behaviour as NetworkPolicy has.

phoracek replied:
And with subnets the default is the opposite? I see that the DHCP example only mentions ports, so I would expect it does not allow any subnets, as no ports attribute would allow no ports?

I'm ok with this if it copies the NetworkPolicy behavior, it is just confusing to me.

rhrazdil replied:
The default is the same, with nothing being allowed by default.

We need a way to express in a policy to
1) not change a subnet/port allow list
2) allow all ports/subnets

ad1) is done by not specifying `subnets` or `ports` array
ad2) is done by specifying `ports: []` or `subnets: []`

```
... Items in this list are combined using a logical OR operation. If this field is 
empty or missing, this rule matches all destinations (traffic not restricted by destination).
If this field is present and contains at least one item, this rule allows traffic only if the
traffic matches at least one item in the to list.
```

The case for empty list is the same as is done in network policies, the difference is that ommiting the key means no-op in this CNI, but means `allow-all` in NetworkPolicy

phoracek replied:
> the difference is that ommiting the key means no-op in this CNI, but means allow-all in NetworkPolicy

Why are we different in that?

phoracek replied:
Nevermind, we don't have to be the same.

rhrazdil replied:
Addressed this in https://github.com/rhrazdil/cidr-filtering-cni/pull/13/commits/c5ac9f84f954769faa83bb0c2fcfbb98ebb0c0b2 (Aligned with NetworkPolicy)

Now both, not specifying a subnet/port field has the same meaning as specifying an empty array.

phoracek commented on README.md:
```diff
@@ -36,80 +36,172 @@ spec:
 
+## API Reference
+
+- ingress:
```
Could you describe defaults and explain how they are merged together?

Commit:
Address review comments

- renamed CNI plugin to bridge-filtering
- addressed review comments

Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>

phoracek commented on README.md:
```diff
@@ -32,84 +40,180 @@ spec:
+metadata:
+  name: deny-all
+  labels:
+    bridge-filtering: ""
+    br1-with-cidr-filtering: ""
```
Could we namespace the NAD reference? `bridge-filtering/br1-with-cidr-filtering: ""`. I think it would make it little clearer what it the purpose of it while still allowing you to filter via kubectl

rhrazdil replied:
I'm not sure if it wouldn't make users think that they may link networkAttachmentDefinition from other namespace?

phoracek replied:
Ha, I see your point. Let's keep it as is then

phoracek commented on README.md:
```diff
@@ -32,84 +40,180 @@ spec:
+> **_NOTE:_**  A ConfigMap may refer to many NetworkAttachmentDefinition specs. The order in which ConfigMaps are process is not defined.
+
+
+To allow a pod to communicate to external entities, let's create the following `ConfigMap`, that allows pod to reach all ports
```
You should mention that you are going to do something about ports 80 and 8080 here as well. Otherwise it is confusing to read the example after this paragraph - especially since you speak about "all ports"

rhrazdil replied:
Yes, removed the `all ports` which is a mistake

phoracek commented on README.md:
```diff
@@ -32,84 +40,180 @@ spec:
+      "egress": {
         "subnets": [
           {
             "subnet": {
```
What's the motivation to have this "subnet" object here?

rhrazdil replied:
Removed

phoracek commented on README.md:
```diff
@@ -32,84 +40,180 @@ spec:
+          {
+            "subnet": {
+              "cidr": "255.255.255.255/32",
+              "except": []
```
Why mentioning it then?

rhrazdil replied:
Removed

phoracek commented on README.md:
```diff
@@ -32,84 +40,180 @@ spec:
+        ]
       },
-      "egress": {}
+      "ingress": {
```
"by default, all subnets are denied"

rhrazdil replied:
But the "policy" configurations are additive, the other configs allow subnets for ingress

phoracek replied:
Oh, I see, thanks.

BTW so subnets and ports are all mixed together? So I cannot say port X on subnet A and port Y on subnet B, without allowing port Y on subnet A?

rhrazdil replied:
This is now changed to the way NetworkPolicies work.
So packet must match both, port and subnet, within singe policy configuration, in order to be accepted.

phoracek replied:
If subnets and ports are coupled now, does that mean you have to explicitly state the subnet in ingrees here?

rhrazdil replied:
Yes, forgot to update the ConfigMaps in README

Commit:
Accept traffic only when matches both subnet and port

defined in a policy.
With this change, packet is accepted only when it matches
both, subnet and port, defined in a particular policy.

Additionally, KUBECONFIG path is propagated from daemonset
env variable.

Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>

Commit:
Align API with NetworkPolicy in case empty array is specified

When specifying
ingress.subnets: []
ingress.ports: []

or ommitting subnets/ports completely, do not allow any port/subnet.

To match all subnets/ports, use:
ingress.subnets:[{}]
of ingress.ports: [{}]

To match all ports on udp/tcp, use:
ingress.ports: [{"protocol": "udp"}]

By ommitting ports key, port number won't be used for matching.

Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>

phoracek commented on README.md:
```diff
@@ -32,84 +40,183 @@ spec:
+is the `bridge-filtering`, that handles configuration of nftables rules on the pod.
+
+By default, all ingress and egress traffic is dropped.
+To allow specific ingress and egress CIDR blocks/ports, create ConfigMaps(s) referencing a NetworkAttachmentDefinition spec.config name in the ConfigMap's label.
```
```suggestion
To allow specific ingress and egress CIDR blocks/ports, create ConfigMap(s) referencing a NetworkAttachmentDefinition spec.config name in the ConfigMap's label.
```

phoracek commented on README.md:
```diff
@@ -32,84 +40,183 @@ spec:
+The `egress.subnets` attribute allows pod to reach any IP address in subnet `192.168.0.0/16`, except for subnet `192.168.150.0/24` and `192.168.151.151`.
+The `egress.ports` attribute allows pod to reach ports 80, and 8080 over TCP protocol.
+
+To allow any webserver responses to reach the pod, create the following `ConfigMap`:
```
```suggestion
To allow any pod requests to reach the webserver, create the following `ConfigMap`:
```

No?

rhrazdil replied:
Actually, no. In both, ingress and egress we match destination port of a packet.
The problem is, client opens connection with a service from a random port, to which is then sent the response.
Normally, this is handled by accepting packets matching "related connection". But the nftable module that enables this feature is not installed by default. 
So for ingress, I need to open all tcp ports to have the response accepted.

Alternatively, we could match source port of ingressing packets, which would probably make more sense without conntrack, but for the price that we deviate from NetworkPolicy
/cc @maiqueb

maiqueb replied:
yeah, afaiu, we don't do conntrack. We probably should.

I actually wasn't aware we required a particular module to have access to it. Could you add more info about this @rhrazdil ?

maiqueb replied:
Another thing that pops to mind: wouldn't the user be expected to provision both ingress / egress in the client to allow the response to go through ? 

*If* so, I think we should document this limitation / requirement we make of them - i.e. dear user, we do not have conntrack. Plan your policies with this in mind.

**If** we eventually decide to implement conntrack, that would change.

phoracek replied:
+1

@rhrazdil let's first evaluate how much extra effort it would require. If it is more than trivial, I'd rather release with the limitation for now

rhrazdil replied:
Using conntrack shoudln't imply a big change to provisioning the rules, I'll just need to figure out why I wasn't able to create rules that use conntrack (maybe load the kernel module?). But, Miguel found that Calico is using conntrack in nftable rules on our nodes and confirmed that conntrack is available on openshift nodes. So if calico can use it, we must be able to use it as well.

For this PR, I would go with documenting the limitation, so that we can proceed. Then get in other open PRs, especially https://github.com/rhrazdil/cidr-filtering-cni/pull/14, and then deliver conntrack.
Is this an acceptable approach?

phoracek replied:
Works for me, let's get this merged o/

rhrazdil replied:
Added the Limitations section to README: https://github.com/rhrazdil/cidr-filtering-cni/pull/13/commits/596c0f9e26f868109539625203024087939d216b

oshoval reviewed (commented):
Thanks

Reviewed the bash only
Looks good
Few nits to consider (only ideas in case you like them)

1. Can we wrap the `ip netns ... || exitWithError`
with a function that do the op and if not succeed prints an error ?
Unless it won't be more readable

```
ip netns exec "$CNI_CONTAINERID" $(nft_add_rule ${table_type} ${NFT_TABLE} "${CNI_IFNAME}"-${config_id}-${direction}-ports ${match_iface} $CNI_IFNAME ip protocol ${protocol,,} counter meta mark set meta mark "|" 0x00010000) || exitWithError "${CNI_VERSION}" "Failed to allow all ports on ipv4 ${protocol} in chain: ${CNI_IFNAME}-${config_id}-${direction}-ports"
```

2. Can we consider some text file with the rules, parameterized, that the script will select which parts to apply, replace parameters and then apply as few chunks as possible ?
Would it make it easier to follow ?

3. Some of the `ip netns ...` don't have exitWithError is it fine ?
why the others do have, after all there is set -eo pipefail

4. If there are terms like `"${CNI_IFNAME}"-${config_id}-${direction}-ports` that repeat, maybe worth to use a var that equal it and then use it all over, so the commands will be easier to grasp ?

rhrazdil committed:
Update README.md

Co-authored-by: Petr Horacek <p.horacek94@gmail.com>

maiqueb commented:
> Thanks
> 
> Reviewed the bash only Looks good 3 nits to consider (only ideas in case you like them)
> 
> 1. Can we wrap the `ip netns ... || exitWithError`
>    with a function that do the op and if not succeed prints an error ?
>    Unless it won't be more readable
> 
> ```
> ip netns exec "$CNI_CONTAINERID" $(nft_add_rule ${table_type} ${NFT_TABLE} "${CNI_IFNAME}"-${config_id}-${direction}-ports ${match_iface} $CNI_IFNAME ip protocol ${protocol,,} counter meta mark set meta mark "|" 0x00010000) || exitWithError "${CNI_VERSION}" "Failed to allow all ports on ipv4 ${protocol} in chain: ${CNI_IFNAME}-${config_id}-${direction}-ports"
> ```
> 
> 2. Can we consider some text file with the rules, parameterized, that the script will select which parts to apply, replace parameters and then apply as few chunks as possible ?
>    Would it make it easier to follow ?

I'm against this. IMO, that's just yet another indirection (that I'm particularly not used to) to follow.

> 3. Some of the `ip netns ...` don't have exitWithError is it fine ?
>    why the others do have, after all there is set -eo pipefail
> 4. If there are terms like `"${CNI_IFNAME}"-${config_id}-${direction}-ports` that repeat, maybe worth to use a var that equal it and then use it all over, so the commands will be easier to grasp ?

Commit:
Wrap ip netns exec with a function

netns_exec_nft, which also handles reporting error,
so that it doesn't have to be repeated on each call.

Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>

Commit:
Check JSON formatting of data in ConfigMap

Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>

rhrazdil commented:
@oshoval Thanks for your review, I have addressed your points 1., 3. and 4.

I'm not convinced that generating a ruleset file would make things considerably better.

oshoval commented on bridge-filtering:
```diff
@@ -0,0 +1,313 @@
+    local direction="$4"
+    local match_addr="saddr"
+    if [[ "$direction" == "egress" ]]; then
+        local match_addr="daddr"
```
No need local here as its already declared local above

rhrazdil replied:
[Yep, thanks](https://github.com/rhrazdil/cidr-filtering-cni/pull/13/commits/c03eb9b9294e1922674c685b4e99cec2df2d5055)

oshoval commented on bridge-filtering:
```diff
@@ -0,0 +1,313 @@
+get_ip_version() {
+    local ip_address="$1"
+    if [[ "$ip_address" =~ .*:.* ]]
+    then
```
I think better to have the then line above please

rhrazdil replied:
[Done](https://github.com/rhrazdil/cidr-filtering-cni/pull/13/commits/c03eb9b9294e1922674c685b4e99cec2df2d5055)

oshoval reviewed (commented):
Nits
beside that lgtm (bash wise)
Thanks

Commit:
Fix indentation and remove extra local keyword

Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>

Commit:
Document limitation of not using conntrack

Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>

maiqueb commented on README.md:
```diff
@@ -1,27 +1,43 @@
 
-See the example below:
+The plugin is currently not using conntrack to allow response traffic. The user is expected to configure ingress/egress policies accordingly.
+That means that for clients, egress configuration may filter desired IP subnets and ports and ingress configuration should allow all traffic or filter by IP subnet
+without filtering by ports.
+Vice versa for servers.
```
I would simplify this sentence - I personally get a bit confused when reading it. I think it has too many words.

How about something along the lines of:
```
Policies applying to clients (in a client/server architecture) must remember to accept the return traffic for their requests.
```
I wouldn't even mention the vice-versa for servers part.

rhrazdil replied:
[Thanks, Done](https://github.com/rhrazdil/cidr-filtering-cni/pull/13/commits/9fb106b692fb73da4518627a34020acbfc01a52b)

maiqueb commented on README.md:
```diff
@@ -1,27 +1,43 @@
+# bridge-filtering
 
-This plugin allows user to define allow list for ingress and egress packet filtering for a Kubernetes pod.
+This CNI plugin allows the user to control traffic flow at the IP address or port level (OSI layer 3 or 4) for particular applications in the Kubernetes cluster, thus specifying how a pod is allowed to communicate with various network "entities". The pod interface must be a port of a bridge (for example a kubevirt virtual machine connected with bridge binding method). By default, all ingress and  ...
+Users may create special `ConfigMaps` in the pod namespace to indicate allowed ingress or egress connections.
+This can be done for network layers L3 and L4. The supported L4 protocols are UDP and TCP.
```
Reads well, but I would add some extra-breathing room. How about (feel free to discard this suggestion ...):

```suggestion
This CNI plugin allows the user to control traffic flow at the IP address or port level (OSI layer 3 or 4) for particular applications in the Kubernetes cluster, thus specifying how a pod is allowed to communicate with various network "entities". 

The pod interface **must** be a port of a bridge (for example a kubevirt virtual machine connected with bridge binding method). By default, all ingress and egress traffic is denied.

Users may create special `ConfigMaps` in the pod namespace to indicate allowed ingress or egress connections.
This can be done for network layers L3 and L4. The supported L4 protocols are UDP and TCP.
```

rhrazdil replied:
[Thanks, Done](https://github.com/rhrazdil/cidr-filtering-cni/pull/13/commits/9fb106b692fb73da4518627a34020acbfc01a52b)

maiqueb commented on README.md:
```diff
@@ -1,27 +1,43 @@
+Users may create special `ConfigMaps` in the pod namespace to indicate allowed ingress or egress connections.
+This can be done for network layers L3 and L4. The supported L4 protocols are UDP and TCP.
+
+Since the nftable rules implementing traffic filtering are created when a pod is being created, the CNI cannot update the provisioned rules if those are updated in the `ConfigMaps`.
```
```suggestion
Since the nftable rules implementing traffic filtering are only applied when a pod is created, the CNI **cannot** update the provisioned rules if those are updated in the `ConfigMaps`.
```

I personally usually like to highlight words like cannot / will not etc. Makes it harder to miss.

rhrazdil replied:
[Thanks, Done](https://github.com/rhrazdil/cidr-filtering-cni/pull/13/commits/9fb106b692fb73da4518627a34020acbfc01a52b)

maiqueb commented on README.md:
```diff
@@ -1,27 +1,43 @@
+This can be done for network layers L3 and L4. The supported L4 protocols are UDP and TCP.
+
+Since the nftable rules implementing traffic filtering are created when a pod is being created, the CNI cannot update the provisioned rules if those are updated in the `ConfigMaps`.
+If a configuration in a `ConfigMap` is changed, all the pods that use it need to be re-created, in order to have up-to-date configuration.
```
```suggestion
If a configuration in a `ConfigMap` is changed, **all the pods that use it need to be re-created**, in order to have up-to-date configuration.
```

rhrazdil replied:
[Thanks, Done](https://github.com/rhrazdil/cidr-filtering-cni/pull/13/commits/9fb106b692fb73da4518627a34020acbfc01a52b)

maiqueb commented on README.md:
```diff
@@ -32,84 +48,186 @@ spec:
+The `example-network` specifies two CNI plugins, the first cni in the above is a `cnv-bridge`, connecting pods to linux-bridge `br1`, and the second
+is the `bridge-filtering`, that handles configuration of nftables rules on the pod.
+
+By default, all ingress and egress traffic is dropped.
```
```suggestion
Once a pod is subject to one of these policies, all ingress and egress traffic is dropped.
```

I honestly don't know if my suggestion is more confusing, or clears matters. Up to you.

rhrazdil replied:
I think it' not 100% correct, as it makes it sound that the default is applied only one a configMap configuration references the NAD, but adding the plugin in the NAD is enough

maiqueb replied:
I'm just trying to avoid the reader misunderstanding that once this plugin is installed "all traffic is dropped" by default.

What if we remove the "one of" from my suggestion ?  I'm just trying to highlight this is what achieves the subsection's goal `Making a pod subject to bridge-filtering policies` . 

... probably redundant. Your call.

rhrazdil replied:
On second thought, I see your point. [Updated](https://github.com/rhrazdil/cidr-filtering-cni/compare/9fb106b692fb73da4518627a34020acbfc01a52b..a7024bded23bd19b8bdd5c1a8efb88b9f49e32eb)

maiqueb commented on README.md:
```diff
@@ -32,84 +48,186 @@ spec:
+To allow specific ingress and egress CIDR blocks/ports, create ConfigMap(s) referencing a NetworkAttachmentDefinition spec.config name in the ConfigMap's label.
+Each ConfigMap additionally needs to have `bridge-filtering` label to be enabled for the bridge-filtering plugin.
+
+The following manifest is an example of a minimal `ConfigMap`, that doesn't yet allow any traffic. Since denying all traffic is done by default, it doesn't needed to exist.
```
```suggestion
The following manifest is an example of a minimal `ConfigMap`, that doesn't yet allow any traffic. This configuration is redundant, since denying all traffic is done by default.
```

rhrazdil replied:
[Thanks, Done](https://github.com/rhrazdil/cidr-filtering-cni/pull/13/commits/9fb106b692fb73da4518627a34020acbfc01a52b)

maiqueb commented on README.md:
```diff
@@ -32,84 +48,186 @@ spec:
+
+By default, all ingress and egress traffic is dropped.
+To allow specific ingress and egress CIDR blocks/ports, create ConfigMap(s) referencing a NetworkAttachmentDefinition spec.config name in the ConfigMap's label.
+Each ConfigMap additionally needs to have `bridge-filtering` label to be enabled for the bridge-filtering plugin.
```
```suggestion
Each ConfigMap additionally needs to have the `bridge-filtering` label to enable the bridge-filtering plugin.
```

rhrazdil replied:
[Thanks, Done](https://github.com/rhrazdil/cidr-filtering-cni/pull/13/commits/9fb106b692fb73da4518627a34020acbfc01a52b)

maiqueb commented on README.md:
```diff
@@ -32,84 +48,186 @@ spec:
+Notice the labels of the ConfigMap.
+The first label, `bridge-filtering` ensures that the `ConfigMap` configuration is collected by the CNI plugin, as
+the plugin only lists `ConfigMaps` with this label.
+The second label, `br1-with-cidr-filtering` refers to the `NetworkAttachmentDefinition` spec name, in the CNI configuration JSON.
```
```suggestion
Notice the labels of the `ConfigMap` specified above:
1. `bridge-filtering` ensures that the `ConfigMap` configuration is collected by the CNI plugin, as the plugin only lists `ConfigMaps` with this label.
2. `br1-with-cidr-filtering` refers to the `NetworkAttachmentDefinition` spec name, in the CNI configuration JSON.
```

rhrazdil replied:
[Thanks, Done](https://github.com/rhrazdil/cidr-filtering-cni/pull/13/commits/9fb106b692fb73da4518627a34020acbfc01a52b)

maiqueb reviewed (commented):
All I have are mainly suggestions on how the improve (opinionated) the text.

Regarding the `Usage` heading, I think adding some sub-sections would make it easier to follow. 

How about something like (poor markdown interpretation follows ....): 
```
## Usage
...
### Making a network subject to network policies
In there you'd show the net-attach-def configuration, where your plugin in added in the plugin chain.
Finish this sub-section w/ the message that once you configure this, all traffic is dropped by default.
### Punching holes in the firewall
Now you show your config maps. I really like how you chose to paint the picture using an example / use case driven. 

Maybe introduce yet another level here, separating ingress from egress. Not sure if it's worth the trouble.
```

Commit:
Improve README wording, emphasize important parts

Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>

phoracek reviewed (approved):
Thanks!

maiqueb reviewed (approved):
Thank you for your patience throughout this review.
